File: cpp/trie/trie.cpp
#include <string.h>

#include "trie.h"

Node::Node(Node *parent=nullptr, char c=0, size_t depth=0)
{
    this->parent = parent;
    this->depth = depth;
    this->c = c;
}

Trie::Trie()
{
    root = new Node();
    string_count = 0;
    current_node = nullptr;
    caching = true;
}

void Trie::clear_cache()
{
    for(const auto &search: cache)
        delete search.second;

    cache.clear();
}

bool Trie::contains(const char *str)
{
    current_node = get_node(str);

    if(current_node)
        if(current_node->is_string) return true;

    return false;
}

void Trie::dfs(Node *node, std::map<Node*, bool> &visited, std::vector<char*> &strings)
{
    visited[node] = true;

    if(node->is_string)
    {
        char *str = trace_string(node);
        if(str) strings.push_back(str);
    }

    for(const auto& [key, n]: node->children)
    {
        if(visited.count(n) == 0) 
            dfs(n, visited, strings);
    }
}

std::vector<char*> Trie::find(const char *str)
{
    if(caching && cache.count(str) > 0)
        return *cache[str];

    current_node = get_node(str);

    if(current_node)
    {
        auto strings = strings_from_node(current_node);
        
        if(caching)
            cache[str] = strings;
        
        return *strings;
    }

    return {};
}

Node *Trie::get_node(const char *str)
{
    current_node = root;
    int len = strlen(str);

    for(int i = 0; i < len; i++)
    {
        if(current_node->children.count(str[i]) > 0)
        {
            current_node = current_node->children[str[i]];
            if(i + 1 == len) return current_node;
        }
        else break;
    }

    return nullptr;
}

void Trie::insert(const char *str)
{
    int len = strlen(str);
    current_node = root;

    for(int i = 0; i < len; i++)
    {
        if(current_node->children.count(str[i]) == 0)
        {
            Node *node = new Node(current_node, str[i], i +1);
            current_node->children[node->c] = node;
        }

        current_node = current_node->children[str[i]];

        if(i + 1 == len)
        {
            current_node->is_string = true;
            string_count++;
        }
    }
}

void Trie::_remove(Node *node, char key)
{
    free(node->children[key]);
    node->children.erase(key);

    if(node->children.size() > 0)
        return;
    
    _remove(node->parent, node->c);
}
        

void Trie::remove(const char *str)
{
    current_node = get_node(str);

    if(current_node && current_node->is_string)
    {
        if(!current_node->children.empty())
            current_node->is_string = false;
        else
            _remove(current_node->parent, current_node->c);

        string_count--;
    }
}

std::vector<char*> *Trie::strings_from_node(Node *node)
{
    auto strings = new std::vector<char*>;

    std::map<Node*, bool> visited;

    dfs(node, visited, *strings);

    return strings;
}

char* Trie::trace_string(Node *node)
{
    current_node = node;

    char *str;
    str = (char*)malloc(sizeof(char) * node->depth + 1);

    if(str)
        str[node->depth] = 0;
    else
        return nullptr;

    for(int i = node->depth; i > 0; i--)
    {
        str[i - 1] = current_node->c;
        current_node = current_node->parent;
    }

    return str;
}

File: js/trie.js
class Node {
    constructor(parent=null, char=null) {
        this.parent = parent;
        this.children = {};
        this.char = char;
        this.isString = false;
    }
}

module.exports = class Trie {
    constructor() {
        this.root = new Node();
        this.cache = {};
        this.stringCount = 0;
        this.caching = true;
    }

    clearCache() {
        for(search in this.cache) {
            delete this.cache[search];
        }
    }

    contains(str) {
        var currentNode = this.getNode(str);

        if(currentNode) {
            if(currentNode.isString) { return true; }
        }

        return false;
    }

    dfs(node, visited, strings) {
        visited.set(node, true);

        if(node.isString) {
            strings.push(this.traceString(node));
        }

        for(var n in node.children) {
            if(!visited.has(node.children[n])) {
                this.dfs(node.children[n], visited, strings);
            }
        }
    }

    find(prefix) {
        if(this.caching && prefix in this.cache) {
            return this.cache[prefix];
        }

        var currentNode = this.getNode(prefix);

        if(currentNode !== null) {
            var strings = this.stringsFromNode(currentNode);
            
            if(this.caching) { this.cache[prefix] = strings; }

            return strings
        }

        return [];
    }

    getNode(str) {
        var currentNode = this.root;
        var len = str.length;

        for(var i = 0; i < len; i++) {
            if(str[i] in currentNode.children) {
                currentNode = currentNode.children[str[i]];
                if(i + 1 === len) { return currentNode; }
            } else { break; }
        }

        return null;
    }

    insert(str) {
        var currentNode = this.root;
        var len = str.length;

        for(var i = 0; i < len; i++) {
            if(!(str[i] in currentNode.children)) {
                var node = new Node(currentNode, str[i]);
                currentNode.children[node.char] = node;
            }

            currentNode = currentNode.children[str[i]];

            if(i + 1 === len) {
                currentNode.isString = true;
                this.stringCount++;
            }
        }
    }

    _remove(node, key) {
        delete node.children[key];

        if(Object.keys(node.children).length > 0 || node.parent === null) {
            return;
        }

        this._remove(node.parent, node.char);
    }

    remove(str) {
        var currentNode = this.getNode(str);

        if(currentNode !== null && currentNode.isString) {
            if(currentNode.children.size > 0) {
                currentNode.isString = false;
            } else {
                this._remove(currentNode.parent, currentNode.char);
            }

            this.stringCount--;
        }
    }

    stringsFromNode(node) {
        var strings = [];
        var visited = new Map();

        this.dfs(node, visited, strings);

        return strings;
    }

    traceString(node) {
        var currentNode = node;
        var str = [];
        
        while(currentNode !== null) {
            str.push(currentNode.char);
            currentNode = currentNode.parent;
        }

        return str.reverse().join('');
    }
}
